Makefile

```
@@ -198,6 +201,10 @@ $(SHARE)/opencv-python: opencv-python/setup.py | $(ACTIVATE_VENV) $(SHARE) $(SHA
 $(BIN)/ocrd: $(SHARE)/opencv-python
 endif
 
+ifeq ($(PYTHON_VERSION),3.10)
+# Python 3.10.x does not work with current kraken.
+override OCRD_MODULES := $(filter-out ocrd_kraken, $(OCRD_MODULES))
+endif
```

This defies several layers of module selection mechanisms that are already in place.

If the user chose to install the ocrd_kraken module, then it is wrong to silently ignore it.

The proper way to remove a module for a certain Python version _by default_ would be to include it in `DISABLED_MODULES`.

---

So

```make
DISABLED_MODULES ?= cor-asv-fst opencv-python ocrd_ocropy
ifeq ($(PYTHON_VERSION),3.10)                            
# Python 3.10.x does not work with current kraken.       
DISABLED_MODULES += ocrd_kraken                          
endif                                                    
```

would still allow overriding with `OCRD_MODULES` but disables it by default for 3.10?

---

Not quite. That would still override any explicit user (or Docker) setting for `DISABLED_MODULES`. But we only want to override the actual default:

```make
ifeq ($(PYTHON_VERSION),3.10)
# Python 3.10.x does not work with current kraken.
DISABLED_MODULES ?= cor-asv-fst opencv-python ocrd_ocropy ocrd_kraken
else
DISABLED_MODULES ?= cor-asv-fst opencv-python ocrd_ocropy
endif
```

---

My patch disables `ocrd_kraken` unconditionally for Python 3.10 (which is necessary because that combination does not work). The proposed alternatives don't do that. Therefore I don't think they are equally good or even better.

---

We could discuss how to handle submodules which are disabled automatically because they don't work. Instead of silently disabling them, there could be a report (for example at the end of the build process) about those disabled submodules.

---

> My patch disables `ocrd_kraken` unconditionally for Python 3.10

exactly.

> (which is necessary because that combination does not work)

thus, let it fail (loudly)!

> Therefore I don't think they are equally good or even better.

As argued above, they are better (you did not address the arguments).

> We could discuss how to handle submodules which are disabled automatically because they don't work. Instead of silently disabling them, there could be a report (for example at the end of the build process) about those disabled submodules.

That would only complicate matters further, and goes against the user expectations of running `make`. If a module does not work in your environment (whatever the cause or kind), then the simplest and uniform answer is that module's build failing (while others can continue if running with `-k`). Also, the build is incremental, so as soon as the user fixed it (by disabling the module or installing by hand), one can continue.